Week3/homework/6-step3.js

```diff
@@ -4,3 +4,4 @@ const arr2d = [[1, 2], [3, 4], [5, 6]];
 const arr3d = [[[1, 2], [3, 4]], [[5, 6], [7, 8]]];
 
 // add your solution here, or add a comment on how you would tackle this problem
+//Need help creating multidimensional arrays.
```

Comment:
Okay, a multi-dimensional array is a fancy word for arrays in arrays.
So, let's say I have a normal array.
```
let arr1 = [1, 2, 3];
```
and I create another array
```
let arr2 = [3 4 5];
```
Both of these arrays are just normal arrays, you run from one end to the other and you have "simple" values (numbers in this case).
Now, what if you put them both in an array:
```
let arr3 = [arr1, arr2]
```
What does this look like?
Well it is just a normal array like any other, but instead of the values being "simple" values (numbers, strings, booleans etc.) the values are another array.  So it looks like this:
```
[arr1, arr2]
```
But, arr1 looks like this: `[1, 2, 3]` 
and arr2 looks like this: `[4, 5, 6]`
so if you get it all expanded arr3 looks like this:
```
[[1, 2, 3], [4, 5, 6]]
```
If you do a for loop over it and print out the results you will get:
```
[1, 2, 3]
[4, 5, 6]
```
So each element of the array instead of being a 1 or 2 or 3 etc. is an array.  So how do you get to the values inside these "inner" arrays?
You have to do exactly the same thing you were doing before (going through the array) but on the inner arrays.

The easiest way to do this sort of thing is called `recursion`.  It is a simple but very useful trick.  I'll create something that will go through and print out the values in the array that are not themselves arrays.
```
function printAllTheItems( arr ) {  
  let values = [];  // This is where we will collect all the values 
  for (let i = 0; i < arr.length; i++){  // typical for loop to traverse all the elements of the array

   // When we come to a new element, we have to ask ourselves what kind of value is it?  Is it another array?
  // If it is, then we need to repeat what we are doing in this function with this "new" array.  So we call the 
  // function again and pass in the current item i.e. arr[i] which is itself an array.
  // in the end our function will return a list of values and we need to catch that and add it to our current values
    if (Array.isArray(arr[i])) {  
      values = values.concat(printAllTheItems(arr[i]));  // setting our values to its current value plus 
                                                                                        //the return from the new call to printAllTheItems.
    } else { // if the current value is not an Array, we just add it to the values array. 
      values = values.concat([arr[i]]);
    }
  }
// Once we have made it through the whole array, including getting the values from the "sub" arrays, we return
// the values array.
  return values; 
};
```
I hope this makes it clearer what is happening.  All we do is go through the array, and every time we run into another array, we call the same function but with the sub array as input, and collect the results that come back.  The cool thing about this is that it will just keep on going from one Array to another and another until it runs out of arrays and collects all the non-array values and returns them.

Calling the above functions works with `arr2d` and `arr3d` and will work on an array that contains lots and lots of sub-arrays.  So it will work on something like this:
```
const arrNd = [[[[1, 2]], [3, 4], [[[5, 6], [7, 8]], [9, 10]], [11, 12]], [[[[[[13, 14]]]]]]];
```
I'm not sure if this explanation has made things better or worse. 😉